### PR TITLE
Preserve highlighted Tree Item when switching between Sources and Out…

### PR DESCRIPTION
**Pull request by nyrosmith**
…line

Associated Issue: #5514

### CODE REVIEW

**Comment by lukaszsobek** [comment]
Looks good to me - highlight stays in place, any edge cases removing the lines might cause?

File: src/components/PrimaryPanes/SourcesTree.js [hunk 1]
```diff
@@ -91,6 +91,19 @@ class SourcesTree extends Component<Props, State> {
...
+  componentDidMount() {
+    const { selectedSource } = this.props;
+    if (selectedSource) {
+      const highlightItems = getDirectories(
+        selectedSource.get("url"),
+        this.state.sourceTree
+      );
+
+      // eslint-disable-next-line react/no-did-mount-set-state
+      this.setState({ highlightItems });
```
Comment (lukaszsobek) [inline src/components/PrimaryPanes/SourcesTree.js]: Could we move this to the constructor?

**Comment by jasonLaster** [comment]
CC @darkwing

File: src/components/PrimaryPanes/SourcesTree.js (hunk 1 above)
Comment (nyrosmith) [reply src/components/PrimaryPanes/SourcesTree.js]: That was my first attempt but it didn't work

Comment (darkwing) [reply src/components/PrimaryPanes/SourcesTree.js]: "react/no-did-mount-set-state" is scaring me; let me think on this

**Comment by darkwing** [comment]
@nyrosmith What do you think about moving the highlightedItems to a parent state like I did here:  

https://github.com/devtools-html/debugger.html/pull/5501

That way we don't need the extra renders that `react/no-did-mount-set-state` brings?

**Comment by nyrosmith** [comment]
Good idea. Let's make it so!

**Comment by darkwing** [comment]
@nyrosmith Would you like to update this or should we?  Thanks!

**Comment by darkwing** [comment]
I'm taking this over, will update soon

**Comment by darkwing** [comment]
Hey @nyrosmith !  I spent a while with this issue today and I shared your frustration.  Your solution here is the easy way but I'm not sure it's the best way.

I started from scratch and *almost* got there.  Here's what I did:

https://gist.github.com/darkwing/8edf5a6e7076b735f423386c80ea373f

Essentially I:

- Calculated the `highlightItems` and `focusedItem` in the `render` method, not relying on state, because that state disappears when you switch to the `Outline` tab

- Used the `props.focused` prop within `ManagedTree` to set the initial `focusedItem`

This solved 90% of the problem but, since I was using the currently selected source in the render method, there was a bit of "jank" when expanding and contracting items.

I thought I'd pass this on for you to look at and see if you can cross the finish line :)

**Comment by darkwing** [comment]
Let's close this for now and pick it up when it has more priority.  Thank you for your effort @nyrosmith !

**Pull request closed without merging**